### PR TITLE
Mediafile pdf presentation with angular-pdf and pdf.js (Fixes #1664).

### LINKED ISSUE #1664: Use pdf.js presentation function in angularjs
We have to migrate our old pdf.js presentation code (from OpenSlides 1.7, mediafile app) to new angularjs client in OpenSlides 2.0 (master). It's one of our last feature which we have to migrate to 2.0 code base.

Here some required work packages:
* Move and integrate current pdf.js files to mediafiles/static/js/
* add the presentation controls (go to next/previous/first/custom slide, toggle fullscreen mode) to new sidebar div container (currently still in openslides/core/static/templates/core/projector-controls.html); buttons should send by ng-click an `update_elements` post request to rest api (see e.g. countdown controls)
* add new slide controller and slide template for presentation div container (see e.g. countdown in core/projector.js)

@boehlke: Can you try it in "your" mediafiles app? Let me know if you need some help.

### PR DESCRIPTION
**Pull request by emanuelschuetze**
Thanks to André Böhlke for contribution core functionality!
Some additional template improvements by Emanuel.

### CODE REVIEW

**Comment by ostcar** [comment]
have not reviewed it. But its ok for me

File: gulpfile.js [hunk 1]
```diff
@@ -80,7 +88,8 @@ gulp.task('translations', function () {
 });
 
 // Gulp default task. Runs all other tasks before.
-gulp.task('default', ['js-libs', 'css-libs', 'fonts-libs', 'ckeditor', 'translations'], function () {});
+gulp.task('default', ['js-libs', 'css-libs', 'fonts-libs', 'ckeditor', 'pdfjs', 'translations'], function () {});
+
```
Comment (normanjaeckel) [inline gulpfile.js]: Remove newline.

File: openslides/core/static/templates/projector.html [hunk 2]
```diff
@@ -7,6 +7,7 @@
 <link rel="stylesheet" href="static/css/projector.css">
 <link rel="icon" href="/static/img/favicon.png">
 <script src="static/js/openslides-libs.js"></script>
+<script src="static/js/openslides-libs.worker.js"></script>
```
Comment (normanjaeckel) [inline openslides/core/static/templates/projector.html]: Please add an issue that we have to find another solution for this. It is not best practice to add extra libs and generate extra requests for this. Maybe we can concat this file together with openslides-libs.js.

File: openslides/mediafiles/static/js/mediafiles/projector.js [hunk 3]
```diff
@@ -2,6 +2,62 @@
...
 'use strict';
 
-angular.module('OpenSlidesApp.mediafiles.projector', ['OpenSlidesApp.mediafiles']);
+angular.module('OpenSlidesApp.mediafiles.projector', ['OpenSlidesApp.mediafiles'])
 
-}());
+.config([
+    'slidesProvider',
+    function(slidesProvider) {
+    slidesProvider.registerSlide('mediafiles/mediafile', {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: Indent these lines.

File: openslides/mediafiles/static/js/mediafiles/projector.js [hunk 4]
```diff
@@ -2,6 +2,62 @@
...
+        }
+
+        $scope.$watch(function() {
+            return $scope.element.scale;
+        }, updateScale);
+
+        updateScale();
+
+        $scope.getNavStyle = function(scroll) {
+            if(scroll > 100) return 'pdf-controls fixed';
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: Is this valid JavaScript? Please use normal style:
```javascript
if () {
    ...
} else {
    ...
}
```

File: openslides/mediafiles/static/js/mediafiles/projector.js [hunk 5]
```diff
@@ -2,6 +2,62 @@
...
+            else return 'pdf-controls';
+        };
+
+        $scope.onError = function(error) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: This function declaration seams useless.

File: openslides/mediafiles/static/js/mediafiles/projector.js [hunk 6]
```diff
@@ -2,6 +2,62 @@
...
+            console.log(error);
+        };
+
+        $scope.onLoad = function() {
+            $scope.loading = '';
+        };
+
+        $scope.onProgress = function(progress) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: This function declaration seams useless.

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 7]
```diff
@@ -111,6 +126,111 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
         $scope.delete = function (mediafile) {
             Mediafile.destroy(mediafile.id);
         };
+
+        // ** PDF presentation functions **/
+        // show document on projector
+        $scope.showPdf = function (mediafile) {
+            var postUrl,
+                data;
+            if($scope.presentedMediafiles.length > 0) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Add whitespace after `if`.

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 8]
```diff
@@ -111,6 +126,111 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
...
+                // update first mediafile, at the moment there should not be more
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: This is not an update but an reset. Are you sure that you want to reset the PDF if calling "showPDF"?

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 9]
```diff
@@ -111,6 +126,111 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
...
+            $http.post('/rest/core/projector/1/update_elements/', postData);
+        }
+
+        function getCurrentlyPresentedMediafile() {
+            return $scope.presentedMediafiles[0];
+        }
+
+        $scope.mediafileGoPrevious = function () {
+            var mediafileElement = getCurrentlyPresentedMediafile();
+            if (mediafileElement.page <= 1) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Proposal: Switch the logic: if ...page > 1 then sendCommand. Thats all. No empty return in the middle of the function neccessary.

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 10]
```diff
@@ -111,6 +126,111 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
...
+                return;
+            }
+            sendMediafileCommand({
+                page: parseInt(mediafileElement.page) - 1
+            });
+        };
+        $scope.mediafileGoNext = function () {
+            var mediafileElement = getCurrentlyPresentedMediafile();
+            if (mediafileElement.page >= mediafileElement.numPages) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Here the same.

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 11]
```diff
@@ -111,6 +126,111 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
...
+        };
+        $scope.mediafileChangePage = function(pageNum) {
+            sendMediafileCommand({
+                pageToDisplay: pageNum
+            });
+        };
+        $scope.mediafileRotate = function () {
+            var rotation;
+            var currentRotation = $scope.mediafile.rotation;
+            if (currentRotation === 0) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Proposal:
```javascript
if (currentRotation === 270 {
    rotation = 0;
} else {
    rotation += 90;
}
```

File: package.json [hunk 12]
```diff
@@ -7,13 +7,14 @@
     "gulp-angular-gettext": "~2.1.0",
     "gulp-concat": "~2.6.0",
     "gulp-if": "~2.0.0",
+    "gulp-jshint": "~2.0.0",
     "gulp-minify-css": "~1.2.3",
+    "gulp-rename": "^1.2.2",
```
Comment (normanjaeckel) [inline package.json]: Please use tilde style.

**Comment by normanjaeckel** [comment]
No time for checking out. Just some coding style and thelike.

File: gulpfile.js (hunk 1 above)
Comment (emanuelschuetze) [inline gulpfile.js]: done

File: openslides/core/static/templates/projector.html (hunk 2 above)
Comment (emanuelschuetze) [inline openslides/core/static/templates/projector.html]: see #1906

File: openslides/mediafiles/static/js/mediafiles/projector.js (hunk 3 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: done

File: openslides/mediafiles/static/js/mediafiles/projector.js (hunk 4 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: done

File: openslides/mediafiles/static/js/mediafiles/projector.js (hunk 5 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: removed function

File: openslides/mediafiles/static/js/mediafiles/projector.js (hunk 6 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/projector.js]: removed function

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 7 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/site.js]: done

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 8 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/site.js]: yes, it's good enough for 2.0b2. we will improve the whole pdf presentation performance later.

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 9 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/site.js]: done

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 10 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/site.js]: done

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 11 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/static/js/mediafiles/site.js]: done

File: package.json (hunk 12 above)
Comment (emanuelschuetze) [inline package.json]: done

**Comment by emanuelschuetze** [comment]
All done. Rebased.

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 8 above)
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Issue?

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 13]
```diff
@@ -111,6 +126,105 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
...
+
+        // ** PDF presentation functions **/
+        // show document on projector
+        $scope.showPdf = function (mediafile) {
+            var postUrl,
+                data;
+            if ($scope.presentedMediafiles.length > 0) {
+                // update first mediafile, at the moment there should not be more
+                var uuid = $scope.presentedMediafiles[0].uuid;
+                postUrl = '/rest/core/projector/1/update_elements/';
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: See #1910.

**Comment by emanuelschuetze** [comment]
Rebased.

File: bower.json [hunk 14]
```diff
@@ -31,9 +31,10 @@
     "ckeditor": "~4.5.4",
     "angular-ckeditor": "~1.0.0",
     "roboto-condensed": "~0.3.0",
-    "open-sans-fontface": "https://github.com/OpenSlides/open-sans.git#1.4.2.post1"
+    "open-sans-fontface": "https://github.com/OpenSlides/open-sans.git#1.4.2.post1",
+    "angular-pdf": "~1.3.0"
   },
   "resolutions": {
-    "angular": "^1.2.x"
+    "angular": ">= 1.4.9"
```
Comment (normanjaeckel) [inline bower.json]: Why not `~1.4.9`?

Comment (emanuelschuetze) [inline bower.json]: ~1.4.9 will match all 1.4.x versions but will miss 1.5.0.
^1.4.9 will match any 1.x.x release including 1.5.0, but will hold off on 2.0.0.
What we need is 1.4.9+ (>=1.4.9).

And yes. We should do `>=1.4.9 <1.5`

Comment (emanuelschuetze) [inline bower.json]: Fixed

**Pull request merged**